pmd-xml/src/main/java/net/sourceforge/pmd/lang/xml/rule/SaxonDomXPathQuery.java

```diff
@@ -0,0 +1,190 @@
+import net.sf.saxon.sxpath.XPathEvaluator;
+import net.sf.saxon.sxpath.XPathExpression;
+import net.sf.saxon.sxpath.XPathStaticContext;
+import net.sf.saxon.sxpath.XPathVariable;
+import net.sf.saxon.trans.XPathException;
+
+final class SaxonDomXPathQuery {
+
+    private static final NamePool NAME_POOL = new NamePool();
+
+    private static final SimpleDataKey<DocumentWrapper> SAXON_DOM_WRAPPER
+        = DataMap.simpleDataKey("pmd.saxon.dom.wrapper");
+
+    /** The XPath expression as a string. */
+    private final String xpath;
+    /** The executable XPath expression. */
+    private final XPathExpressionWithProperties xpathExpression;
+
+
+    private final Configuration configuration;
+
+    SaxonDomXPathQuery(String xpath, String defaultNsUri, List<PropertyDescriptor<?>> properties) {
+        this.xpath = xpath;
+        configuration = new Configuration();
+        configuration.setNamePool(NAME_POOL);
+        xpathExpression = makeXPathExpression(this.xpath, defaultNsUri, properties);
+    }
+
+    private XPathExpressionWithProperties makeXPathExpression(String xpath, String defaultUri, List<PropertyDescriptor<?>> properties) {
+        final IndependentContext xpathStaticContext = new IndependentContext(configuration);
+        xpathStaticContext.declareNamespace("fn", NamespaceConstant.FN);
+        xpathStaticContext.setDefaultElementNamespace(defaultUri);
+
+
+        // Register PMD functions
+        Initializer.initialize(xpathStaticContext);
+
+        Map<PropertyDescriptor<?>, XPathVariable> xpathVariables = declareXPathVariables(properties, xpathStaticContext);
+
+        try {
+            final XPathEvaluator xpathEvaluator = new XPathEvaluator(configuration);
+            xpathEvaluator.setStaticContext(xpathStaticContext);
+            XPathExpression expression = xpathEvaluator.createExpression(xpath);
+            return new XPathExpressionWithProperties(
+                expression,
+                xpathVariables
+            );
+        } catch (final XPathException e) {
+            throw new ContextedRuntimeException(e)
+                .addContextValue("XPath", xpath);
+        }
+    }
+
+    private Map<PropertyDescriptor<?>, XPathVariable> declareXPathVariables(List<PropertyDescriptor<?>> accessibleProperties, XPathStaticContext xpathStaticContext) {
+        Map<PropertyDescriptor<?>, XPathVariable> xpathVariables = new HashMap<>();
+        for (final PropertyDescriptor<?> propertyDescriptor : accessibleProperties) {
+            final String name = propertyDescriptor.name();
+            if (!isExcludedProperty(name)) {
+                final XPathVariable xpathVariable = xpathStaticContext.declareVariable(null, name);
+                xpathVariables.put(propertyDescriptor, xpathVariable);
+            }
+        }
+        return Collections.unmodifiableMap(xpathVariables);
+    }
+
+    private boolean isExcludedProperty(String name) {
+        return "xpath".equals(name)
```

Comment:
should we exclude here `defaultNsUri` as well?

Reply:
Indeed, thanks!